internal/mempool/mempool.go

```diff
@@ -1159,36 +1288,60 @@ func (g *txDescGraph) insert(txDesc *mining.TxDesc, findTx txDescFind) {
-func (mv *txMiningView) updateStatsDescendantsAdded(txDesc *mining.TxDesc) {
-	txHash := txDesc.Tx.Hash()
-	txSize := int64(txDesc.Tx.MsgTx().SerializeSize())
-	seen := make(map[chainhash.Hash]struct{})
-	mv.txGraph.forEachDescendant(txHash, seen, func(descendant *mining.TxDesc) {
-		descendantStats := mv.ancestorStats[*descendant.Tx.Hash()]
-		descendantStats.Fees += txDesc.Fee
-		descendantStats.SizeBytes += txSize
-		descendantStats.TotalSigOps += txDesc.TotalSigOps
-		descendantStats.NumAncestors++
-	})
+func (mv *txMiningView) updateStatsDescendantsAdded(baseTxDesc *mining.TxDesc) {
+	txHash := baseTxDesc.Tx.Hash()
+	_, baseTxHasStats := mv.ancestorStats[*txHash]
+	seenDescendants := make(map[chainhash.Hash]struct{},
+		ancestorTrackingLimit+1)
+
+	mv.txGraph.forEachDescendantPreOrder(txHash, seenDescendants,
+		func(descendant *mining.TxDesc) bool {
+			descendantTxHash := *descendant.Tx.Hash()
+			_, descendantTxHasStats := mv.ancestorStats[descendantTxHash]
+			if !descendantTxHasStats {
+				// Cannot update ancestor stats for a transaction or its
+				// descendants if it does not have ancestor stats tracked.
+				return false
+			}
+
+			if !baseTxHasStats {
+				// If the base tx does not have stats, then remove stats for all
+				// of its descendants.
+				// This can happen during a reorg when two graphs of
+				// transactions get connected through the base transaction.
+				delete(mv.ancestorStats, descendantTxHash)
+				return true
+			}
+
+			// Recalculate the stats for descendants of a transaction added
+			// back to the mempool.
+			// Although this does walk ancestors of a descendant
+			// the complexity is bound by the ancestorTrackingLimit.
+			// There is room for additional optimizations such as directly
+			// adding the the base transaction's ancestor stats to all
+			// descendants if there is no potential for double-counting
+			// the base transaction's ancestors' stats, though this scenario
+			// seems unlikely enough to not do this.
+			mv.maybeUpdateAncestorStats(descendant)
+			return true
+		})
 }
 
 // addTransaction inserts a TxDesc into the mining view if it has a parent
 // or child in the view, or has a relationship with another transaction through
 // findTx.
-func (mv *txMiningView) addTransaction(txDesc *mining.TxDesc, findTx txDescFind) {
+func (mv *txMiningView) addTransaction(txDesc *mining.TxDesc, findTx txDescFind, isNew bool) {
 	// Track the txn in the graph and attempt to relate it to at least one other
 	// transaction.
 	mv.txGraph.insert(txDesc, findTx)
 
 	if mv.trackAncestorStats {
-		// Ensure that the statistics for this transaction are up to date
-		// given the current ancestors in the view.
-		mv.updateBundleStats(txDesc)
-
-		// If a transaction is added back to the view,
-		// reconnect it with transactions already in the view.
-		// This is necessary to account for reorgs.
-		mv.updateStatsDescendantsAdded(txDesc)
+		mv.maybeUpdateAncestorStats(txDesc)
+		if !isNew {
```

Comment:
This can be expressed as `len(mv.txGraph.childrenOf[*txDesc.Tx.Hash()]) > 0`, which simplifies the method signature. Only a `!isNew` transaction can have descendants that are already in the main pool.